cincinnati/src/plugins/internal/graph_builder/release.rs

```diff
@@ -80,70 +81,89 @@ pub fn create_graph(releases: Vec<Release>) -> Result<cincinnati::Graph, failure
-
-            previous.iter().try_for_each(|version| {
-                let previous = match graph.find_by_version(&version.to_string()) {
-                    Some(id) => id,
-                    None => graph.add_release(cincinnati::Release::Abstract(
-                        cincinnati::AbstractRelease {
-                            version: version.to_string(),
-                        },
-                    ))?,
-                };
-                graph.add_edge(&previous, &current).and_then(|edge| {
-                    trace!(
-                        "Adding EdgeIndex({}): from {} to {:?}",
-                        edge.index(),
-                        &version.to_string(),
-                        current
-                    );
-                    Ok(())
-                })
-            })?;
-
-            next.iter().try_for_each(|version| {
-                let next = match graph.find_by_version(&version.to_string()) {
-                    Some(id) => id,
-                    None => graph.add_release(cincinnati::Release::Abstract(
-                        cincinnati::AbstractRelease {
-                            version: version.to_string(),
-                        },
-                    ))?,
-                };
-                graph.add_edge(&current, &next).and_then(|edge| {
-                    trace!(
-                        "Adding EdgeIndex({}): from {:?} to {}",
-                        edge.index(),
-                        current,
-                        &version.to_string(),
-                    );
+                .try_for_each(|version| -> Fallible<()> {
+                    let next = match graph.find_by_version(&version.to_string()) {
+                        Some(id) => id,
+                        None => {
+                            warn!("Adding abstract release for {}", version.to_string());
+                            graph.add_release(cincinnati::Release::Abstract(
+                                cincinnati::AbstractRelease {
+                                    version: version.to_string(),
+                                },
+                            ))?
+                        }
+                    };
+                    graph.add_edge(&current, &next)?;
+
                     Ok(())
                 })
-            })
         })?;
 
     Ok(graph)
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn create_graph_tolerates_nonexistent_edges() -> Fallible<()> {
+        let releases = vec![Release {
+            source: "test-0.0.1".to_string(),
+            metadata: Metadata {
+                kind: MetadataKind::V0,
+                version: semver::Version::from((0, 0, 1)),
+                next: Default::default(),
+                previous: vec![semver::Version::from((0, 0, 0))],
+                metadata: Default::default(),
+            },
+        }];
+
+        let mut graph = create_graph(releases).unwrap();
+
+        assert_eq!(graph.prune_abstract(), 1);
```

Comment:
Why do we need to call `graph.prune_abstract()`? We do not use it in the `pub fn create_graph` so using it here is confusing.

Reply:
`Graph::prune_abstract` prunes and returns the number of abstract releases in the graph. The test uses this to assert that there was indeed 1 abstract release there. I found this the easiest way to ensure the currently expected behavior.

Reply:
We are not using the method anywhere in the code because we currently don't explicitly prune abstracts releases from the graph. Due to their lack of metadata they are implicitly filtered out by plugins such as the channel filter or the architecture filter.

Reply:
Got it, thanks.